test/dkv-dagger-hashimoto-test.js

```diff
@@ -189,4 +194,181 @@ describe("Basic Func Test", function () {
+        if (arr.length === 0) {
+          results.push(memo.concat(cur));
+        }
+        permute(arr.slice(), memo.concat(cur));
+        arr.splice(i, 0, cur[0]);
+      }
+  
+      return results;
+    }
+  
+    return permute(inputArr);
+  }
+
+  function random_generator() {
+    const numbers = Array((1 << (shardSizeBits - maxKvSizeBits))-1).fill().map((_, index) => index + 1);
+    numbers.sort(() => Math.random() - 0.5);
+    return numbers.slice(0,randomChecks);
+  }
+
+  it("runs full mining cycle", async function () {
+    const [owner, addr1, addr2] = await ethers.getSigners();
+    let wallet = ethers.Wallet.createRandom().connect(owner.provider);
+    const SystemContract = await ethers.getContractFactory("TestSystemContractDaggerHashimoto");
+    const sc = await SystemContract.deploy();
+
+    const MinabledKV = await ethers.getContractFactory("TestDecentralizedKVDaggerHashimoto");
+    // 64 bytes per data, 16 entries in shard, 4 random access
+    const kv = await MinabledKV.deploy(
+      [maxKvSizeBits, shardSizeBits, randomChecks, minimumDiff, 
+       targetIntervalSec, cutoff, diffAdjDivisor, coinbaseShare, sc.address],
+      0, //startTime
+      0, //storageCost
+      0, // dcfFactor
+      ethers.utils.formatBytes32String("") //genesisHash
+    );
+    await kv.deployed();
+
+    /* Debug : check the meta at beginning */
+    if(verbose){
+      let mining_debug_last_shard = async function(str){
+        const lastKvIdx = await kv.lastMinableShardIdx();
+        const info = await kv.infos(lastKvIdx);
+        console.log("Stage %s MiningInfo :",str);
+        console.log(info);
+      };
+      mining_debug_last_shard("INIT");
+    }
+    let l = 0;
+    let dataList = [];
+    for (let i = 0; i < (1 << (shardSizeBits - maxKvSizeBits)); i++) {
+      let d = "0x";
+      for (let j = 0; j < (1<<maxKvSizeBits) / 32; j++) {
+        d = ethers.utils.hexConcat([d, ethers.utils.keccak256(hexlify4(l))]);
+        l = l + 1;
+      }
+      dataList.push(d);
+      // PUT won't change the diff info
+      await kv.put(local_idx_2_key(i), dataList[i]);
+    }
+
+    const mineTs = 120;
+    kv.setTimestamp(240);
+    const meta = await kv.calculateDiffAndInitHash(0,1,mineTs);
+    const diff = meta.diff.toNumber();
+    const nonce = 1;
+
+    let count = 0;
+    let maskedData = [];
+    while (count < LIMIT)
+    {
+      if(verbose){
+        console.log("Turn %d ", count);
+      }
+      const a = ethers.utils.hexlify(meta.hash0);
+      const b = padLeft32(ethers.utils.hexlify(wallet.address));
+      const c = padLeft32(ethers.utils.hexlify(ethers.BigNumber.from(mineTs)));
+      const d = padLeft32(ethers.utils.hexlify(ethers.BigNumber.from(nonce)));
+      const h0 = keccak256(ethers.utils.concat([a,b,c,d]));
+      /* random fashion: ready hard to find */
+      //const randomList = random_generator();
+      const randomList = [count % (1 << (shardSizeBits - maxKvSizeBits))];
```

Comment:
To be perfectly honest, I am not quite sure. Myself kinda questions the real usage of nonce.
Simply looking into this case, I have to try different combination of data slice to "collide" a good one which can pass Hashimoto check. Say one shard have n trunks, and randomness check needs m data slice to submit as proof. What I really want to do, is do a combinatorics m out of n, then do permutation on those m slices, it is C(n,m) * m! . Considering  m is big enough, perhaps I'd better use random algorithm rather than brute force, that's the PoW comes from. Theoretically, if I were miner, I'd like to keep nonce fixed, instead of moving around. So you can see here I always keep that fixed.
But I would like to hear from Qi. I guess I missed some parts. @qizhou